### ISSUE TITLE
[bug] upgraded to 4.6.0 and plugins behind SSL apache proxy now fail... ?

### ISSUE THREAD

**Issue by fschaer** [grafana/grafana#9729]
- What Grafana version are you using?
4.6.0
- What datasource are you using?
graphite v0.9
- What OS are you running grafana on?
CentOS 7.4, apache https proxy with X509 user certs
- What did you do?
just upgraded
- What was the expected result?
The plugins should work
- What happened instead?
The plugins now fail to load giving a pop up failure saying for instance :

`
NetworkError when attempting to fetch resource. Instantiating https://x.y.z:20005/grafana/public/plugins/grafana-piechart-panel/module.js Loading plugins/grafana-piechart-panel/module
`
(nothing else is displayed, and I could not find that message anywhere in the logs)
I tried another plugin (the clock) and it shows the same behaviour.

Looking in firefox in the developper tabs, the only thing I can see is a request to https://x.y.z:20005/grafana/public/plugins/grafana-piechart-panel/module.js failing

The error reported by firefox is (translated from french) :

`An error hapenned while connecting to x.y.z:20005. The SSL pair failed to negociate acceptable security params. error code : SSL_ERROR_HANDSHAKE_FAILURE_ALERT `

In apache set to debug mode, I see the folowing :

```
[Mon Oct 30 13:12:34.609910 2017] [ssl:info] [pid 14630] [client 1.2.3.4:55742] AH01964: Connection to child 3 established (server x.y.z:443)
[Mon Oct 30 13:12:34.610177 2017] [ssl:debug] [pid 14630] ssl_engine_kernel.c(1890): [client 1.2.3.4:55742] AH02043: SSL virtual host for servername x.y.z found
[Mon Oct 30 13:12:34.626503 2017] [ssl:info] [pid 14630] [client 1.2.3.4:55742] AH02008: SSL library error 1 in handshake (server x.y.z:443)
[Mon Oct 30 13:12:34.626647 2017] [ssl:info] [pid 14630] SSL Library Error: error:140890C7:SSL routines:ssl3_get_client_certificate:peer did not return a certificate -- No CAs known to server for verification?
[Mon Oct 30 13:12:34.626665 2017] [ssl:info] [pid 14630] [client 1.2.3.4:55742] AH01998: Connection closed to child 3 with abortive shutdown (server x.y.z:443)
```

Nothing in the grafana logs even in debug mode
Looks like the user cert is not passed to the https proxy (anymore ?).

I tried to revert to previous grafana version and everything works again.
I tried to check the "skip TLS verification" box in the datasource but thid did not change anything.
Not sure what else I can try to work this around and pinpoint the now-failing part...

Any ideas ?

**Comment by bergquist** [comment]
Strange. The skip TLS verification for data sources is for outgoing requests. This seems to be about incoming requests. What version did you use before upgrading to 4.6.0?

**Comment by fschaer** [comment]
Hi,

I(re)tried this morning with grafana-4.5.2-1.x86_64 which worked .

I'm currently trying to exclude the /grafana/public url from HTTPS authentication requirement as a possible workaround suggested in other bug reports, but for now I'm/it's still failing ;)

**Comment by bergquist** [comment]
I guess thats the issue.

How is the connection between apache and grafana configured downstream? Is apache terminating SSL?

**Comment by fschaer** [comment]
Yes, apache is terminating SSL and redirecting using a proxypass to the same host on port 3000.
But before that, I have a <Location "/"> with a require that restricts access to only some users of some CAs.

But for now, adding another location section does not prevent the error from hapenning, or at least with this one either in front or after my X509 requirements :

`
  <Location "/grafana/public/">
    Require all granted
  </Location>
`

I'll see if I can move the SSLVerifyClient inside the location where it's required instead of having all the ssl options defined in the whole vhost, maybe that's my error in the work around... I say workaround, because it was working before and not anymore - and I'd be happy if I could really restrict who has access to even (my) public resources :/

**Comment by fschaer** [comment]
OK, got this working like this:

In the location section matching / : set SSLVerifyClient require (I had this set right at the end of the https proxy vhost)
Below that location, add a new location with this :

`  <Location "/grafana/public/">
    Require all granted
    SSLVerifyClient none
  </Location>
`

So, indeed, this works around the issue. But this also means that now, everyone in the world can read /grafana/public resources, and I'm not very enthousiast about this... I'd be much happier if access to public resources could still pass the user authentication (certificate) if it exists, as it seemed to work with previous versions :/

**Comment by bergquist** [comment]
Hmm

Exposing frontend publicly shouldn't matter in a security perspective in Grafana since no data is transferred that way. But this doesn't feel like a good enough solution.

**Comment by PaulKuiper** [comment]
I observe the same problem on our test servers. We wil try your solution, but indeed it is not vey nice.

**Comment by mattbostock** [comment]
@fschaer: Are you using self-signed X509 certificates?

Does disabling TLS verification using the following setting work?

```
[plugins]
app_tls_skip_verify_insecure = true
```

This setting is used here:
https://github.com/grafana/grafana/pull/9378/files#diff-b0758c1824aa8af08525a9e7a781b386R25

Disabling TLS verification should be avoided, although this was the implicit behaviour in Grafana versions prior to 4.6.0. @bergquist It seems there might be a need for custom TLS certs for proxying plugin requests?

**Comment by torkelo** [comment]
I think this might be solved by https://github.com/grafana/grafana/commit/e19b4a92919c4d0cfbe0739fde22b1c394d47fbc

**Comment by bergquist** [comment]
This should be fixed in 4.6.1 (released within hours) If not. let's reopen this issue
